### PR TITLE
Debugger tensor runner

### PR DESCRIPTION
**Pull request by jindrahelcl**
Hi, have a look at this piece of code and tell me your ideas.. It introduces some kind of debugger runner which can run arbitrary number of tensors from graph and store them in a data series (which is later saved in an NPZ file). 
There were/are two larger problems with this:
1. the runner API forces too many things, one of them (not unreasonable) being that the output of a runner should be a batch-sized array. Hence I needed to add the `batch_dim_*` parameters so I could
transpose it. But the problem remains for tensors that do not have any batch dimension (e.g. train_mode) but I suppose those tensors are not very interesting to write out. 
2. we do not have any writer interface so I had to make an ugly hack to learning utils.

Note this will need some more work done before it can be merged, I just wanted to discuss if this is the good way for us to start doing some model inspection via exporting tensor/activations..

See the `bpe.ini` test for usage.

This is the commit message:

The runner runs a bunch of tensors (specified either by name or a
reference) and generates an output series of format

```
(batch, {tensor_name: value(tensor_shape\batch_dim)})
```

that means batch of dicts indexed by tensor name. The batch
dim from the tensors has been transposed out to the outer list axis.

### CODE REVIEW

**Comment by jindrahelcl** [comment]
Motivation is:
1. this can help us further with creating ways of how to look at a trained model
2. also it can help us comparing our model with others, e.g. during writing importing scripts

**Comment by varisd** [comment]
I think it might be useful to support more than one execution result (so we can also debug ensembles). I guess we can prefix the output tensor with "model_%n", "session_%n" or something similar.

File: neuralmonkey/learning_utils.py [hunk 1]
```diff
@@ -421,6 +421,15 @@ def run_on_dataset(tf_manager: TensorFlowManager,
                 if isinstance(data, np.ndarray):
                     np.save(path, data)
                     log("Result saved as numpy array to '{}'".format(path))
+                elif (isinstance(data, list)
```
Comment (jlibovicky) [inline neuralmonkey/learning_utils.py]: Tohle IHMO nestačí, chtělo by to ještě kontrolovat, že je do dict nampájů.

Comment (jindrahelcl) [reply neuralmonkey/learning_utils.py]: to bych musel projíždět celej ten dict.. Nejradši bych na to použil `typeguard.check_type`, ale byl by to ugly hack, asi jako tohle. Koukni do komentáře v komitu, co jsem k tomu udělal, tam je example

Comment (varisd) [reply neuralmonkey/learning_utils.py]: Mozna by stalo za to, checknout https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/util/nest.py
Ale nevim, jestli se to hodi zrovna na tento problem (nezkoumal jsem do detailu, co vsechno to podporuje)

Comment (jindrahelcl) [reply neuralmonkey/learning_utils.py]: To je užitečný na procházení těch nested struktur, který se používaj třeba pro feed dict nebo pro fetches.. Tady chci jen čekovat typ...

**Comment by jindrahelcl** [comment]
Ten RepresentationRunner se asi pak dá udělat jako syntactic sugar na tenhle TensorRunner.

**Comment by varisd** [comment]
Jsem za rozsireni RepresentationRunneru s tim, ze soucasny repre. runner bude special case.

**Comment by jindrahelcl** [comment]
Tak jsem to všecko udělal.

1. `RepresentationRunner` dědí od `TensorRunner`u a maličko se mu změnilo API (z `used_session: int = 0` na `select_session: int = None`.
2. `TensorRunner` jsem rozšířil o parametr `select_session: int = None`, kterým se volí session. Pokud se nezvolí a je jich víc, je výstupní formát struktura o dimenzích `([batch], [session], {tensor})`, kde hranatý jsou listový a složený jsou dictový. Pokud je jen jedna session nebo pokud se zvolí, pak je to jen `([batch], {tensor})`
3. v `learning_utils` jsem nakonec hodil to kontrolování zneužitím typeguardu. Snad vás to nepohoršuje.

NOTE: nejsou testy na případ s víc sessions.

**Comment by jlibovicky** [comment]
Co brání tohle dodělat a zmergovat?

**Comment by varisd** [comment]
Chybejici rebase.

File: neuralmonkey/runners/tensor_runner.py [hunk 2]
```diff
@@ -43,52 +64,129 @@ def collect_results(self, results: List[Dict]) -> None:
...
                  tensors_by_ref: List[tf.Tensor],
                  batch_dims_by_name: List[int],
                  batch_dims_by_ref: List[int],
-                 output_series: str) -> None:
+                 select_session: int = None) -> None:
+        """Construct a new ``TensorRunner`` object.
+
+        Args:
+            output_series: The name of the generated output data series.
+            toplevel_modelpart: A ``ModelPart`` object that is used as the
```
[the remark this answers is not in the document]
Comment (jindrahelcl) [reply neuralmonkey/runners/tensor_runner.py]: Když je tam dáváš jménem tak může být v runneru pozdě.

**Comment by jindrahelcl** [comment]
Chybějící approve

File: neuralmonkey/runners/tensor_runner.py [hunk 3]
```diff
@@ -43,52 +64,129 @@ def collect_results(self, results: List[Dict]) -> None:
...
+    def __init__(self,
+                 output_series: str,
+                 encoder: ModelPart,
+                 attribute: str = "output",
+                 select_session: int = None) -> None:
+        """Initialize the representation runner.
+
+        Args:
+            output_series: Name of the output series with vectors.
+            encoder: The encoder to use.
```
[the remark this answers is not in the document]
Comment (jindrahelcl) [reply neuralmonkey/runners/tensor_runner.py]: done

File: neuralmonkey/runners/tensor_runner.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jlibovicky) [reply neuralmonkey/runners/tensor_runner.py]: Rozumim. Nechceš teda ještě do toho docstringu dopsat pohádku o tom, jak je všechno lazy a proto není v obecnosti zaručený, že to bude včas vytvořený a proto tam musí být tyhle divné atributy?
